Document breaking changes and stabilizations for uv 0.12 (#20745)
astral-sh/uv: CHANGELOG.md

<!-- prettier-ignore-start -->

## 0.12.0

Since we released uv [0.11.0](https://github.com/astral-sh/uv/releases/tag/0.11.0) in March, we've
accumulated changes that improve correctness, safety, and compatibility, but could break some
workflows. This release contains those changes; many have been marked as breaking out of an
abundance of caution. We expect most users to be able to upgrade without making changes.

This release also stabilizes preview behavior for project discovery, virtual environment safety,

> publishing, Conda environments, and source distributions.

Author: I think we also need a top-level callout for the `uv init` package default change matching some of our commentary in the docs and prior notes about why we reverted the behavior forever ago in 0.3 or something.

The [`uv_build` build backend](https://docs.astral.sh/uv/concepts/build-backend/) now rewrites
`pyproject.toml` in source distributions, as described below. If you have an upper bound in your

> `[build-system]` table, you should update it, e.g., from `<0.12` to `<0.13`.

Author: I think this needs a little more detail? It's not clear how "now rewrites" is compatible with "bump your upper bound?

### Breaking changes

- **Reject unsupported source distribution and wheel archive formats**
  ([#18927](https://github.com/astral-sh/uv/pull/18927))

  [PEP 625](https://peps.python.org/pep-0625/) requires source distributions to use `.tar.gz`
  archives. Previously, uv also accepted legacy formats such as `.tar.bz2`, `.tar.xz`, and

>   `.tar.zst`. Those formats are now rejected, including when referenced by an existing lockfile.

Author: We're not planning to drop zst support, are we?

Author: Should we include a note about this reducing the compression stack attack surface?

  Legacy `.zip` source distributions remain supported for backwards compatibility.

  Wheels and other ZIP archives can no longer contain entries compressed with bzip2, LZMA, or XZ.
  Entries must use the stored, DEFLATE, or zstd compression methods.

  You cannot opt out of this behavior. Rebuild source distributions as `.tar.gz` archives and
  wheels with a supported ZIP compression method. If an existing `uv.lock` references an
  unsupported source distribution, update the dependency and regenerate the lockfile.

- **Prefer stable releases before falling back to pre-releases**
  ([#19993](https://github.com/astral-sh/uv/pull/19993))

>   uv previously allowed pre-releases by default when a direct dependency explicitly requested one

Author: I think some more details about the motivation for this and history of it, e.g., difference from pip etc. our previous resolution mode. etc.

How can this description include no reference to the prerelease modes?

  or a package only published pre-releases. A pre-release requested by a transitive dependency,
  e.g., `example>=2.0.0b1`, could cause resolution to fail even when that pre-release satisfied
  every requirement.

  uv now prefers stable candidates and considers pre-releases when no stable candidate satisfies
  the active constraints. This allows transitive pre-release requirements, but can change which
  versions are selected when both stable and pre-release candidates are available.

  You can opt out of automatic pre-release selection with `--prerelease disallow`. Alternatively,
  `--prerelease allow` considers pre-releases without first preferring stable releases. If you
  explicitly configured `if-necessary-or-explicit`, replace it with `if-necessary`; the old name

>   remains available as a deprecated alias.

Author: Let's expand this last sentence a bit and explai why it's been deprecated or something? Note it will be removed in the future?

- **Respect `--require-hashes` directives in `requirements.txt`**
  ([#19336](https://github.com/astral-sh/uv/pull/19336))

  Previously, `uv pip install` and `uv pip sync` warned about `--require-hashes` inside a
  `requirements.txt` file but still installed dependencies without checking their hashes. Now, the
  directive enables hash-checking mode, just as if `--require-hashes` had been passed on the command
  line.

  For example, this requirements file is no longer accepted because the requirement is neither
  pinned nor hashed:

  ```text
  --require-hashes
  anyio
  ```

  You cannot opt out while the directive is present. Pin every requirement with `==` and provide
  its hash, or remove `--require-hashes` if hash checking is not intended.

> - **Require package entries in `pylock.toml` files**

Author: As a meta note, let's group these with other changes related to `pylock.toml`

  version `1.01.0` should be named `example-1.1.0-py3-none-any.whl`, not
  `example-1.01.0-py3-none-any.whl`.

  Previously, `uv publish` warned about non-normalized filenames but still attempted to upload
  them. It now skips the affected wheels and source distributions instead.

  You cannot opt out of this behavior. Rebuild distributions with normalized filenames before
  publishing.

- **Rewrite `pyproject.toml` files in source distributions for TOML 1.0 compatibility**
  ([#20225](https://github.com/astral-sh/uv/pull/20225))

  Python's `tomllib` only supports TOML 1.0 through Python 3.14, so source distributions using
  TOML 1.1 syntax can fail when built by older frontends. `uv_build` now always rewrites the
  distributed `pyproject.toml` as TOML 1.0 and includes the original as `pyproject.toml.orig`.

  Previously, the rewrite occurred only when TOML 1.1 syntax was detected or the corresponding
  preview feature was enabled. Now, even a TOML 1.0 input is normalized and the source distribution
  gains an additional `pyproject.toml.orig` file.

  You cannot opt out of this behavior. If you compare source distributions byte-for-byte or depend

>   on their exact file list, update those expectations.

Author: lmao "update those expectations", let's not say that....

  Before Python upgrades were supported, `uv python install 3.12 --reinstall` doubled as a way to
  install the latest Python 3.12 patch release. Now that `--upgrade` is available, `--reinstall`
  reinstalls the matching patch releases that are already present.

  For example, if Python 3.12.6 and 3.12.7 are installed, `uv python install 3.12 --reinstall`
  reinstalls both versions instead of installing the latest available 3.12 release.

  You can recover the previous upgrade behavior with `uv python install 3.12 --upgrade`. Combine
  `--upgrade --reinstall` to reinstall only the latest patch.

- **Require `--upgrade-group` to name an existing dependency group**
  ([#18957](https://github.com/astral-sh/uv/pull/18957))

  Previously, `uv lock --upgrade-group docs` silently succeeded even if no `docs` dependency group
  existed. uv now validates the requested group against the project, its workspace members, and
  workspace-level dependency groups.

  You cannot opt out of this behavior. Correct the group name or add it to `[dependency-groups]`.
  Legacy `tool.uv.dev-dependencies` still satisfies `--upgrade-group dev`, and `--frozen` continues

>   to ignore upgrade requests.

Author: I don't think we need this note

<!-- prettier-ignore-start -->

## 0.12.0

Since we released uv [0.11.0](https://github.com/astral-sh/uv/releases/tag/0.11.0) in March, we've
accumulated changes that improve correctness, safety, and compatibility with specifications, but
could break some workflows. This release contains those changes; many have been marked as breaking

> out of an abundance of caution. We expect most users to be able to upgrade without making changes.

Reviewer: The release notes are long, so I'd bold **We expect most users to be able to upgrade without making changes.** as a tl;dr.

<!-- prettier-ignore-start -->

## 0.12.0

Since we released uv [0.11.0](https://github.com/astral-sh/uv/releases/tag/0.11.0) in March, we've

> accumulated changes that improve correctness, safety, and compatibility, but could break some

Author: Maybe?
```suggestion
accumulated changes that improve correctness, safety, and compatibility with specifications, but could break some
```

workflows. This release contains those changes; many have been marked as breaking out of an
abundance of caution. We expect most users to be able to upgrade without making changes.

> The most visible change is to `uv init`, which now creates a packaged project by default. When uv's

Author: Don't frame it as "the most visible", I don't like that.

> project interface stabilized in 0.3, the default application layout was kept intentionally simple:

Author: No? We reverted this due to user confusion while using hatchling and built our own build backend? Did you invent this context?

flat, unpackaged application because requiring a build backend confused users whose projects did
not need to be distributed. Since then, uv has introduced its own build backend, `uv_build`, which
integrates with uv and provides a simpler packaging experience. New projects now use a `src` layout
and `uv_build`, making their code importable, testable, and distributable. The previous flat layout

> remains available with `uv init --no-package` or `uv init --app`.

Author: ```suggestion
Projects created with `uv init` now declare a build system and are packaged by default. This was the default project layout all the way back in v0.3, but we found that the use of the `hatchling` build system was confusing to newcomers and consequently dropped use of a build system by default v0.4. Since then, we've created our own build system (`uv_build`) with tight integration with uv and are excited to restore the default to a best-practice project layout.
```

This release also stabilizes preview behavior for project discovery, virtual environment safety,

> publishing, Conda environments, source distributions, and Unix resource limits.

Author: Let's just drop this paragraph

The [`uv_build` build backend](https://docs.astral.sh/uv/concepts/build-backend/) follows uv's
breaking-release versioning policy. If your `[build-system]` table includes an upper bound such as
`uv_build>=0.11.32,<0.12`, it will continue selecting an older build backend after upgrading uv.
Update that upper bound to `<0.13` to use `uv_build` 0.12. The build backend also changes how

> `pyproject.toml` is included in source distributions, as described below.

Author: Let's drop this last sentence?

The first sentence is confusing, what does that mean?

### Breaking changes

> - **Create packaged projects by default with `uv init`**

Author: Could we generally adjust this in a way that aligns with my docs review change that recently merged?

  ([#19197](https://github.com/astral-sh/uv/pull/19197))

>   Previously, `uv init example` created an unpackaged application containing `main.py` and a

Author: ```suggestion
  Previously, `uv init example` created an unpackaged layout containing `main.py` and a
```

  `pyproject.toml` without a build system. The project could declare dependencies but was not itself
  installed into its virtual environment.

  Now, `uv init example` creates a packaged application with source code in `src/example`, a
  `[build-system]` using `uv_build`, and a `[project.scripts]` entry named `example`. The project
  can be imported from tests or other code, installed as a dependency, and run as a command:

  ```console
  $ uv init example
  $ cd example
  $ uv run example
  Hello from example!
  ```

  Existing projects are unaffected. You can opt out of the new layout with

>   `uv init --no-package example` or `uv init --app example`, which create the previous flat

Author: The `--app` bit will need changing after https://github.com/astral-sh/uv/pull/20773

  application layout.

- **Reject unsupported source distribution and wheel archive formats**
  ([#18927](https://github.com/astral-sh/uv/pull/18927))

  [PEP 625](https://peps.python.org/pep-0625/) requires source distributions to use `.tar.gz`
  archives. Previously, uv also accepted legacy formats such as `.tar.bz2` and `.tar.xz`. Those
  formats are now rejected, including when referenced by an existing lockfile. Legacy `.zip` source
  distributions remain supported for backwards compatibility.

  Wheels and other ZIP archives can no longer contain entries compressed with bzip2, LZMA, or XZ.
  Entries must use the stored, DEFLATE, or zstd compression methods. zstd remains supported for
  wheels and other archives; `.tar.zst` is not accepted specifically as a source distribution
  because PEP 625 requires `.tar.gz`.

  Removing support for uncommon compression methods reduces uv's compression dependencies and the
  attack surface exposed when processing untrusted packages.

  You cannot opt out of this behavior. Rebuild source distributions as `.tar.gz` archives and
  wheels with a supported ZIP compression method. If an existing `uv.lock` references an

>   unsupported source distribution, update the dependency and regenerate the lockfile.

Author: @woodruffw for your review

Reviewer 2: :shipit:

- **Prefer stable releases before falling back to pre-releases**
  ([#19993](https://github.com/astral-sh/uv/pull/19993))

>   Pre-releases are difficult to resolve because requirements are discovered incrementally. uv

Author: ```suggestion
  Pre-releases are difficult to model during resolution because requirements are discovered incrementally. uv
```

  previously required each package's pre-release eligibility to be known before resolution began:
  the default `if-necessary-or-explicit` mode allowed them for direct requirements that explicitly
  requested a pre-release, or for packages that only published pre-releases.

  This meant that a pre-release requirement discovered in a dependency's metadata, e.g.,

>   `example>=2.0.0b1`, could fail to resolve even when a compatible pre-release existed. Users had

Author: "would"?

Author: Can we also avoid saying "Users had to" here? I try not to refer to "Users" this document is for them!

  to add that dependency as a direct requirement or allow pre-releases across their entire
  dependency graph.

  The default mode is now `if-necessary`. uv tries stable candidates first and falls back to
  pre-releases when no stable candidate satisfies the active constraints, including constraints
  discovered transitively. This matches modern pip's pre-release handling, but can select different

>   versions than previous uv releases when stable and pre-release candidates are both available.

Author: ```suggestion
  versions than previous uv releases when both stable and pre-release candidates are available.
```

  You can opt out of automatic pre-release selection with `--prerelease disallow`. Alternatively,
  `--prerelease allow` considers pre-releases without first preferring stable releases, while

>   `--prerelease explicit` only allows them for direct requirements that mention a pre-release.

Author: ```suggestion
  `--prerelease allow` considers pre-releases without first preferring stable releases, and
  `--prerelease explicit` only allows them for direct requirements that mention a pre-release.
```

  The old `if-necessary-or-explicit` mode distinguished between explicitly requested pre-releases
  and packages with no stable releases. That distinction is unnecessary now that `if-necessary`
  handles both cases, including transitive requirements. The old name remains available as an alias

>   but emits a deprecation warning and will be removed in a future release. Replace

Author: ```suggestion
  but is deprecated and will be removed in a future release. Replace
```

  `--prerelease if-necessary-or-explicit` with `--prerelease if-necessary`, and update any

>   corresponding `prerelease` configuration values.

Author: I think we can omit this sentence

  ([#20225](https://github.com/astral-sh/uv/pull/20225))

  uv previously warned when `--project` referred to a missing directory or a file other than
  `pyproject.toml`, but then attempted to continue. This could produce confusing errors later or run
  against an unintended project.

  Now, `uv run --project missing python` fails immediately instead of continuing. You cannot opt
  out of this behavior. Create the directory first or select an existing project. Passing
  `--project path/to/pyproject.toml` remains supported and selects the file's parent directory.

- **Skip distributions with non-normalized filenames when publishing**
  ([#20225](https://github.com/astral-sh/uv/pull/20225))

  Distribution filenames must use normalized package names and versions. For example, a wheel for
  version `1.01.0` should be named `example-1.1.0-py3-none-any.whl`, not
  `example-1.01.0-py3-none-any.whl`.

  Previously, `uv publish` warned about non-normalized filenames but still attempted to upload
  them. It now skips the affected wheels and source distributions instead.

  You cannot opt out of this behavior. Rebuild distributions with normalized filenames before
  publishing.

> - **Rewrite `pyproject.toml` files in source distributions for TOML 1.0 compatibility**

Author: We should move this up, this is an important change

  ([#20225](https://github.com/astral-sh/uv/pull/20225))

  Python's `tomllib` only supports TOML 1.0 through Python 3.14, so source distributions using
  TOML 1.1 syntax can fail when built by older frontends. `uv_build` now always rewrites the
  distributed `pyproject.toml` as TOML 1.0 and includes the original as `pyproject.toml.orig`.

  Previously, the rewrite occurred only when TOML 1.1 syntax was detected or the corresponding
  preview feature was enabled. Now, even a TOML 1.0 input is normalized and the source distribution
  gains an additional `pyproject.toml.orig` file.

  You cannot opt out of this behavior. If a tool needs the original contents or formatting, read

>   `pyproject.toml.orig` from the source distribution instead.

Author: cc @konstin for review

  the latest PyPy 3.10 release, but older bzip2-only patch releases are omitted.

  You cannot opt out of this behavior. Request a newer PyPy patch release instead.

- **Omit excluded-package comments when annotations are disabled**
  ([#20085](https://github.com/astral-sh/uv/pull/20085))

  `uv pip compile --no-annotate` suppresses comments describing the generated requirements file.
  Previously, a footer listing packages excluded with `--unsafe-package` was still included, even
  though annotations were disabled. That footer is now omitted.

  You can recover the footer by removing `--no-annotate`.

### Stabilizations

- **Packaged projects created by `uv init`**
  ([#19197](https://github.com/astral-sh/uv/pull/19197))

  `uv init` now creates an importable, distributable project with a `src` layout, the `uv_build`
  backend, and a command-line entry point by default. Existing projects are unaffected, and
  `uv init --no-package` or `uv init --app` retains the previous flat application layout.

  See the
  [project creation documentation](https://docs.astral.sh/uv/concepts/projects/init/#packaged-applications)

>   for more details.

Author: We already mention this in the breaking changes, we don't need to do so again.

We can just add a note in that section saying...

"This is the stabilization of the `...` feature"? and do that for other stabilized breaking changes?

  `uv init` now rejects `--project`, which is intended to select an existing project. Use a
  positional path to create a project in another directory, or `--directory` to change the working
  directory before initialization.

- **Normalized distribution filenames in `uv publish`**
  ([#20225](https://github.com/astral-sh/uv/pull/20225))

  `uv publish` now skips wheels and source distributions whose package names or versions are not
  normalized, preventing invalid artifacts from being uploaded to package indexes.

- **Conda environments named `base` and `root`**
  ([#20225](https://github.com/astral-sh/uv/pull/20225))

  uv now determines whether a Conda environment is the base environment from its location instead
  of treating `base` and `root` as reserved names. Child environments with either name can be
  discovered and used normally.

- **TOML 1.0-compatible source distributions**
  ([#20225](https://github.com/astral-sh/uv/pull/20225))

  `uv_build` now writes a TOML 1.0-compatible `pyproject.toml` when building source distributions,
  allowing older Python build frontends to consume projects that use newer TOML syntax. The

>   original project file remains available in the archive as `pyproject.toml.orig`.

Author: This makes sense as a stabilization too

- **Automatic open-file limit adjustment on Unix**
  ([#20225](https://github.com/astral-sh/uv/pull/20225))

  On Linux and macOS, uv now raises the soft open-file limit at startup toward the hard limit,
  capped at 1,048,576 descriptors. The higher limit is inherited by subprocesses and helps prevent

>   open-file exhaustion during large dependency installations and builds.

Author: This might be the only non-breaking stabilization?

Can this fail?

Reviewer 2: I don't think so. I guess the syscall that raises the open FD cap itself could fail, but I would expect our testing to reveal that (and FWICT we turn that error into a warning).

Reviewer 2: https://github.com/astral-sh/uv/blob/cc148b8f4b9f75c8514208022e718d50d92f98d3/crates/uv/src/lib.rs#L512-L517

  ```console
  $ uv init example
  $ cd example
  $ uv run example
  Hello from example!
  ```

  Existing projects are unaffected. Use `uv init --no-package example` to create the previous
  unpackaged layout without a build system.

  See the
  [project creation documentation](https://docs.astral.sh/uv/concepts/projects/init/#applications)
  for more details.

  This stabilizes the `packaged-init` preview feature.

- **Reject unsupported source distribution and wheel archive formats**
  ([#18927](https://github.com/astral-sh/uv/pull/18927))

  [PEP 625](https://peps.python.org/pep-0625/) requires source distributions to use `.tar.gz`
  archives. Previously, uv also accepted legacy formats such as `.tar.bz2` and `.tar.xz`. Those
  formats are now rejected, including when referenced by an existing lockfile. Legacy `.zip` source
  distributions remain supported for backwards compatibility.

  Wheels and other ZIP archives can no longer contain entries compressed with bzip2, LZMA, or XZ.
  Entries must use the stored, DEFLATE, or zstd compression methods. zstd remains supported for
  wheels and other archives; `.tar.zst` is not accepted specifically as a source distribution

>   because PEP 625 requires `.tar.gz`.

Reviewer 2: Nit: I think the part about `.tar.zst` at the end is kind of a red herring, I'd just leave it out:

```suggestion
  Wheels and other ZIP archives can no longer contain entries compressed with bzip2, LZMA, or XZ.
  Entries must use the stored, DEFLATE, or zstd compression methods.
```

  Removing support for uncommon compression methods reduces uv's compression dependencies and the
  attack surface exposed when processing untrusted packages.

  You cannot opt out of this behavior. Rebuild source distributions as `.tar.gz` archives and
  wheels with a supported ZIP compression method. If an existing `uv.lock` references an

>   unsupported source distribution, update the dependency and regenerate the lockfile.

Reviewer 2: ```suggestion
  You cannot opt out of this behavior. If you depend on a legacy source
  distribution that uses an unsupported format, we recommend rebuilding
  it as a `.tar.gz` archive and regenerating any lockfile containing
  references to the legacy archive.
```

<!-- prettier-ignore-start -->

## 0.12.0

Since we released uv [0.11.0](https://github.com/astral-sh/uv/releases/tag/0.11.0) in March, we've
accumulated changes that improve correctness, safety, and compatibility with specifications, but
could break some workflows. This release contains those changes; many have been marked as breaking
out of an abundance of caution.

**We expect most users to be able to upgrade without making changes.**

Projects created with `uv init` now declare a build system and are packaged by default. This was
the default project layout all the way back in v0.3, but we found that the use of the `hatchling`
build system was confusing to newcomers and consequently dropped use of a build system by default
in v0.4. Since then, we've created our own build system (`uv_build`) with tight integration with

> uv and are excited to restore the default to a best-practice project layout.

Author: Maybe let's just move this commentary into the breaking changes, since this is the first one we discuss. Let's consolidate this framing with the breaking change entry for this.

There are no breaking changes to the configuration of the
[uv build backend](https://docs.astral.sh/uv/concepts/build-backend/). If your `[build-system]`
table includes an upper bound on `uv_build`, update it to allow `uv_build` 0.12:

```toml
[build-system]
requires = ["uv_build>=0.11.32,<0.12"]
build-backend = "uv_build"
```

becomes:

```toml
[build-system]
requires = ["uv_build>=0.11.32,<0.13"]
build-backend = "uv_build"

> ```

Author: This is a bit too verbose. We don't need this full example.

Reviewer 3: I didn't mind the copypastable example, but at least let's keep the `requires = ...` line?

### Breaking changes

- **Define build systems by default with `uv init`**
  ([#19197](https://github.com/astral-sh/uv/pull/19197))

  Previously, `uv init example` created an unpackaged layout containing `main.py` and a
  `pyproject.toml` without a build system. The project could declare dependencies but was not itself
  installed into its virtual environment.

  Now, `uv init example` defines a `[build-system]` using `uv_build`, places application source
  code in `src/example`, and includes a `[project.scripts]` entry named `example`. Defining a
  build system allows the project to be imported from tests or other code, installed as a
  dependency, and run as a command:

  ```console
  $ uv init example
  $ cd example
  $ uv run example
  Hello from example!
  ```

>   Existing projects are unaffected. Use `uv init --no-package example` to create the previous flat

Reviewer 3: nit:
```suggestion
  Existing projects are unaffected. Use `uv init --no-package example` to create the previous unpackaged
```
